### ISSUE TITLE
Is this a bug, or intended behaviour?

### ISSUE THREAD

**Issue by jesseleite** [wellle/targets.vim#218]
I just installed targets and love the extra operators!  However, it changes behaviour of default operators.  For example, without targets when I `ci(` here:

![2018-11-08 09 29 15](https://user-images.githubusercontent.com/5187394/48204969-32386d80-e339-11e8-996f-907fda9b6ac6.gif)

And with targets in the same situation:

![2018-11-08 09 30 44](https://user-images.githubusercontent.com/5187394/48204990-45e3d400-e339-11e8-97a3-19f784b97ef6.gif)

I appreciate having the ability to `cin(`, but with `ci(` I was actually looking for default vim behaviour.  Is this a bug, or intended behaviour?

**Comment by wellle** [comment]
@jesseleite: Thanks!

Quoting from the Readme:

> ### g:targets_seekRanges
> 
> Default:
> 
> ```vim
> let g:targets_seekRanges = 'cc cr cb cB lc ac Ac lr rr ll lb ar ab lB Ar aB Ab AB rb al rB Al bb aa bB Aa BB AA'
> ```
> 
> Defines a priority ordered, space separated list of range types which can be
> used to customize seeking behavior.
> 
> The default setting generally prefers targets around the cursor, with one
> exception: **If the target around the cursor is not contained in the current
> cursor line, but the next or last target are, then prefer those.** Targets
> beginning or ending on the cursor are preferred over everything else.
> 
> Some other useful example settings:
> 
> Prefer multiline targets around cursor over distant targets within cursor line:
> ```vim
> let g:targets_seekRanges = 'cc cr cb cB lc ac Ac lr lb ar ab lB Ar aB Ab AB rr ll rb al rB Al bb aa bB Aa BB AA'
> ```

So yes, it is intended. You can try the other setting above which should behave like default Vim.

(I went for the different behaviour to keep seeking available in such cases. Often when working within such big blocks I found it annoying to operate on those instead of the more local ones to the right of the cursor in the cursor line.)

**Comment by jesseleite** [comment]
Is it possible to disable seeking altogether on native text objects?  For example, preserve `ci(`, and for me to `cin(` if I want it?

**Comment by wellle** [comment]
Yes, you can disable seeking completely by setting these seek ranges:
```
let g:targets_seekRanges = 'cc cr cb cB lc ac Ac lr lb ar ab lB Ar aB Ab AB'
```
But this applies to all text objects, not only the native ones. And I don't understand why you'd want to disable seeking completely. Is it important for you that `ci(` fails when you're not in a block?

I'd suggest you try this setting first:
```
let g:targets_seekRanges = 'cc cr cb cB lc ac Ac lr lb ar ab lB Ar aB Ab AB rr ll rb al rB Al bb aa bB Aa BB AA'
```
This seeks if there's no text object around the cursor, but works like the native text objects if there is one.

**Comment by jesseleite** [comment]
>Is it important for you that ci( fails when you're not in a block?

Mainly just that it makes certain targets easier to hit, and other (native vim) targets that I rely on harder to hit.  For example, with the default seek ranges, how would I target the parent parenthesis from the cursor in my first gif?

I would also (respectfully) counter your question and ask:  If some users don't want automatic seeking, and are happy using `n` or `l` (ie. `cin(` or `cil(`) to explicitly target distant code from cursor, what is the disadvantage?

Also regarding your first example:

```
let g:targets_seekRanges = 'cc cr cb cB lc ac Ac lr lb ar ab lB Ar aB Ab AB'
```

>But this applies to all text objects, not only the native ones.

Would I be correct in saying all text objects without the use of `n` or `l` would behave as they would without targets.vim, but could still be explicitly targeted with `n` or `l`?

Btw, please don't take this as negative.  This looks like an extremely well designed plugin, just trying to find the right config for me ☺️

**Comment by jesseleite** [comment]
>Is it important for you that ci( fails when you're not in a block?

Mainly just that it makes certain targets easier to hit, and other (native vim) targets that I rely on harder to hit.  For example, with the default seek ranges, how would I target the parent parenthesis from the cursor in my first gif?

![2018-11-08 09 29 15](https://user-images.githubusercontent.com/5187394/48204969-32386d80-e339-11e8-996f-907fda9b6ac6.gif)

I would also (respectfully) counter your question and ask:  If some users don't want automatic seeking, and are happy using `n` or `l` (ie. `cin(` or `cil(`) to explicitly target distant code from cursor, what is the disadvantage?

On that note, regarding this:

```
let g:targets_seekRanges = 'cc cr cb cB lc ac Ac lr lb ar ab lB Ar aB Ab AB'
```

>But this applies to all text objects, not only the native ones.

Would I be correct in saying all text objects without the use of `n` or `l` would behave as they would without targets.vim, but could still be explicitly targeted with `n` or `l`?

Btw, please don't take this as negative.  This looks like an extremely well designed plugin, just trying to find the right config for me ☺️

**Comment by wellle** [comment]
> For example, with the default seek ranges, how would I target the parent parenthesis from the cursor in this gif?

With default settings, it's indeed not possible to select the outer block from that cursor position. I think I could implement it so that `c1i(` would do that. But not sure if that's good enough to be worth implementing it.

> I would also (respectfully) counter your question and ask: If some users don't want automatic seeking, and are happy using `n` or `l` (ie. `cin(` or `cil(`) to explicitly target distant code from cursor, what is the disadvantage?

Users who don't want the seeking can disable it. Interestingly the built in quote text objects `ci"` do seek to the right (but not to the left), but block text objects `ci(` never seek. So the motivation for me was to add seeking for block text objects as well and also allow seeking to the left too. If you prefer to be explicit with `cin"` or `cil(` that's totally fine.

> Would I be correct in saying all text objects without the use of `n` or `l` would behave as they would without targets.vim, but could still be explicitly targeted with `n` or `l`?

Yes, with those settings `ci)` should behave like the native text objects, while you can still use `cin)`.

(But again, the other settings suggested above also work like native text objects in all cases where they do work (inside of a block for example). Only in the cases where they don't find one, seeking would be active.)

> Btw, please don't take this as negative. This looks like an extremely well designed plugin, just trying to find the right config for me ☺️

No worries, I appreciate your feedback, I don't often get it so explicitly, so keep it coming 👍

**Comment by jesseleite** [comment]
>But again, the other settings suggested above also work like native text objects in all cases where they do work (inside of a block for example). Only in the cases where they don't find one, seeking would be active.

Yeah just playing with that setting, and I like it so far!  Still more seeking, but if there's a parent, respect that first.  It'd force me to be more explicit with `n` or `l` in certain situations, but IMO is possibly a best of both worlds?

>No worries, I appreciate your feedback, I don't often get it so explicitly, so keep it coming 👍

That's the problem with open source; Users act entitled 😅  It's free, you owe us nothing!  ...I do appreciate the help though ❤️

Anyway, feel free to close this issue!

**Comment by jesseleite** [comment]
>In general I'm not opposed to make that the default setting, but personally I find the current default less confusing. I might expect ci( to seek into the block I see just right of the cursor, but if I happen to be inside of a bigger block it will act on the outer block.

In a bigger block with more fluff around the cursor line, I 100% agree with you.  More often than not, I would probably be wanting to target something on the cursor line anyway.  So where do you draw the line for a default setting?  I'm not sure honestly 😅 

That said, I think what bothers me most about the default setting is what you say here:

>With default settings, it's indeed not possible to select the outer block from that cursor position.

The default setting is actually making it _**impossible**_ to target something that a user _**could**_ target in native vim.  With the setting you led me to, I have access to both targets, at the expense of needing to be explicit with `n` and `l` in some cases.  That said, I'm not sure I'm willing to part with the parent as a target, when I just installed targets.vim today and noticed the issue so soon.

**Comment by wellle** [comment]
It's good to see this pointed out so clearly. It does indeed seem weird to have this become impossible by the choice of that default setting. I can't promise anything, but I'll keep it in mind and ponder a bit on it 👍

**Comment by jesseleite** [comment]
Cool man!  Appreciate your work on this awesome plugin again though.  I [got the setting in there now](https://github.com/jesseleite/dotfiles/commit/ceb76d53e859035dbb50eaf061c7acff2518713a), so thanks again 🙏
